accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,500 @@
+
+#### Authoring side approach
+
+To distribute a NuGet cross platform plugin as a .NET Tool, plugin authors need to follow these steps:
+
+- Follow [NuGet cross platform plugins](https://learn.microsoft.com/nuget/reference/extensibility/nuget-cross-platform-plugins) guidance.
+- Ensure that the .NET Tool command name begins with `nuget-plugin-*`.
+- Execute `dotnet pack` to generate the package.
+
+```xml
+  <PackAsTool>true</PackAsTool>  
+```
+
+> Note that if the `ToolCommandName` property value is set for a .NET tool and it conflicts with an existing command from another tool, the `dotnet tool install` command will fail.
+
+#### Consumer side approach
+
+The proposed workflow for repositories that access private NuGet feeds, such as Azure DevOps, would be:
+
+1. Ensure that the .NET SDK is installed.
+1. Execute the command `dotnet tool install -g Microsoft.CredentialProvider`.
+1. Run `dotnet restore --interactive` with a private endpoint. It should 'just work', meaning the credential providers installed in step 2 are used during credential acquisition and are used to authenticate against private endpoints.
+
+Upon installation, these global .NET tools are added to the PATH. This allows NuGet to easily determine which file in the package should be run at runtime.
+
+### Security considerations
+
+The [.NET SDK docs](https://learn.microsoft.com/dotnet/core/tools/global-tools#check-the-author-and-statistics) clearly state, `.NET tools run in full trust. Don't install a .NET tool unless you trust the author`.
+This is an important consideration for plugin customers when installing NuGet plugins via .NET Tools in the future.
+
+### Technical explanation
+
+#### Plugin discovery
+
+Currently, plugins are discovered through a convention-based directory structure, such as the `%userprofile%/.nuget/plugins` folder on `Windows`.
+For CI/CD scenarios, and for power users, environment variables can be used to override this behavior.
+Note that only absolute paths are allowed when using these environment variables.
+
+- `NUGET_NETFX_PLUGIN_PATHS`: Defines the plugins used by the .NET Framework-based tooling (NuGet.exe/MSBuild.exe/Visual Studio). This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_NETCORE_PLUGIN_PATHS`: Defines the plugins used by the .NET Core-based tooling (dotnet.exe).
+This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_PLUGIN_PATHS`: Defines the plugins used for the NuGet process, with priority preserved.
+If this environment variable is set, it overrides the convention-based discovery.
+It is ignored if either of the framework-specific variables is specified.
+
+I propose the addition of a `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable.
+This variable will define the plugins, installed as .NET tools, to be used by both .NET Framework and .NET Core tooling.
+It will take precedence over `NUGET_PLUGIN_PATHS`.
```

Comment:
If we're scanning the path, why do we need anything that's plugin specific?

Can we just say we're scannign the path, and if it happens to be a plugin, great.

Reply:
> I don't understand how NUGET_DOTNET_TOOLS+PLUGIN_PATHS is different to NUGET_PLUGIN_PATHS. Does NUGET_PLUGIN_PATHS still look for the netfx and netcore subdirectories?

> If we're scanning the path, why do we need anything that's plugin specific?

Initially, I had the exact same idea to leverage the existing `NUGET_PLUGIN_PATHS` environment variable because it is meant to work in both .NET Framework and .NET Core runtimes, which is perfect for this scenario. Upon reviewing the usage of this environment variable on [GitHub](https://github.com/search?q=NUGET_PLUGIN_PATHS&type=code), I believe we can indeed leverage the existing environment variable. The only reason I added the new environment variable with the name "DOTNET_TOOLS" is to encourage customers to use .NET Tool for developing and distributing the plugin. Additionally, it allows customers to install .NET tools as tool path tools instead of global tools. @dsplaisted asked similar question offline. If team members think that we can leverage the existing `NUGET_PLUGIN_PATHS` environment variable, I am happy to update the spec.